service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java

```diff
@@ -0,0 +1,135 @@
+          if (maybeLock.isPresent()) {
+            // If the lock is present, another thread/instance is getting retrieving a key.
+            // Wait for it to finish and get the key from the DB.
+            log.info("Lock {} is present, waiting for it to expire.", newLock.getLockName());
+            throw new DistributedLockException(
+                String.format(
+                    "Encountered lock %s for user %s",
+                    newLock.getLockName(), linkedAccount.getUserId()));
+          } else {
+            log.info("Lock {} is not present, creating a new lock.", newLock.getLockName());
+            // Lock is not present.
+            // Create a lock and retrieve a key from the fence provider.
+            try {
+              distributedLockDAO.insertDistributedLock(newLock);
+            } catch (DataAccessException e) {
+              log.warn(
+                  String.format(
+                      "Failed to insert lock %s, another thread is doing the same thing.",
+                      newLock.getLockName()));
+              throw new DistributedLockException(
+                  String.format(
+                      "Encountered lock %s for user %s",
+                      newLock.getLockName(), linkedAccount.getUserId()));
+            }
+            var fenceAccountKey = retrieveFenceAccountKey(linkedAccount);
+            fenceAccountKeyService.upsertFenceAccountKey(fenceAccountKey);
+            distributedLockDAO.deleteDistributedLock(
+                newLock.getLockName(), linkedAccount.getUserId());
+            return Optional.of(fenceAccountKey);
+          }
+        });
+  }
+
+  private FenceAccountKey retrieveFenceAccountKey(LinkedAccount linkedAccount) {
+    if (linkedAccount.getId().isEmpty()) {
+      throw new IllegalArgumentException(
+          "Cannot retrieved Fence Account Key for an unsaved Linked Account");
+    }
+    var providerClient = providerOAuthClientCache.getProviderClient(linkedAccount.getProvider());
+    var providerProperties = externalCredsConfig.getProviderProperties(linkedAccount.getProvider());
+    var accessToken =
+        oAuth2Service.authorizeWithRefreshToken(
+            providerClient, new OAuth2RefreshToken(linkedAccount.getRefreshToken(), null));
+    var keyEndpoint = providerProperties.getKeyEndpoint();
+    if (keyEndpoint.isEmpty()) {
+      throw new IllegalArgumentException(
+          "Provider " + linkedAccount.getProvider() + " does not have a key endpoint");
+    }
+    WebClient.ResponseSpec response =
+        WebClient.create(keyEndpoint.get())
+            .post()
+            .header("Authorization", "Bearer " + accessToken.getAccessToken().getTokenValue())
+            .retrieve();
+    String responseBody =
+        response
+            .onStatus(HttpStatusCode::isError, clientResponse -> Mono.empty())
+            .bodyToMono(String.class)
+            .block(Duration.of(11, ChronoUnit.SECONDS));
```

Comment:
Maybe this should be a configurable value

Reply:
Probably. I'll have to wait until the provider config is merged to do that though.